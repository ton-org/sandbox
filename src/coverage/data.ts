import type {Step, TraceInfo} from "ton-assembly-dev-test/dist/trace"
import {Cell} from "@ton/core"

export type Coverage = {
    readonly code: Cell;
    readonly lines: readonly Line[];
};

export type Line = {
    readonly line: string
    readonly info: Covered | Uncovered | Skipped
}

export type Covered = {
    readonly $: "Covered"
    readonly hits: number
    readonly gasCosts: readonly number[]
}

export type Uncovered = {
    readonly $: "Uncovered"
}

export type Skipped = {
    readonly $: "Skipped"
}

export type InstructionStat = {
    readonly name: string
    readonly totalGas: number
    readonly totalHits: number
    readonly avgGas: number
}

export type CoverageSummary = {
    readonly totalLines: number
    readonly coveredLines: number
    readonly uncoveredLines: number
    readonly coveragePercentage: number
    readonly totalGas: number
    readonly totalHits: number
    readonly instructionStats: readonly InstructionStat[]
}

export const buildLineInfo = (trace: TraceInfo, asm: string): readonly Line[] => {
    const lines = asm.split("\n")

    const perLineSteps: Map<number, Step[]> = new Map()

    for (const step of trace.steps) {
        if (step.loc === undefined) continue
        const line = step.loc.line

        perLineSteps.set(line + 1, [...(perLineSteps.get(line + 1) ?? []), step])

        if (step.loc.otherLines.length > 0) {
            for (const otherLine of step.loc.otherLines) {
                perLineSteps.set(otherLine + 1, [...(perLineSteps.get(otherLine + 1) ?? []), step])
            }
        }
    }

    return lines.map((line, idx): Line => {
        const info = perLineSteps.get(idx + 1)
        if (info) {
            const gasInfo = info.map(step => normalizeGas(step.gasCost))

            return {
                line,
                info: {
                    $: "Covered",
                    hits: gasInfo.length,
                    gasCosts: gasInfo,
                },
            }
        }

        if (!isExecutableLine(line)) {
            return {
                line,
                info: {
                    $: "Skipped",
                },
            }
        }

        return {
            line,
            info: {
                $: "Uncovered",
            },
        }
    })
}

const normalizeGas = (gas: number): number => {
    if (gas > 10000) {
        return 26
    }
    return gas
}

export const isExecutableLine = (line: string): boolean => {
    const trimmed = line.trim()
    return (
        !trimmed.includes("=>") && // dictionary
        trimmed !== "}" && // close braces
        trimmed !== "]" && // close bracket
        trimmed.length > 0
    )
}

export const generateCoverageSummary = (
    coverage: Coverage,
): CoverageSummary => {
    const lines = coverage.lines;
    const totalExecutableLines = lines.filter(line => isExecutableLine(line.line)).length

    const coveredLines = lines.filter(
        line => isExecutableLine(line.line) && line.info.$ === "Covered",
    ).length
    const uncoveredLines = totalExecutableLines - coveredLines
    const coveragePercentage = totalExecutableLines === 0 ? 0 : (coveredLines / totalExecutableLines) * 100

    let totalGas = 0
    let totalHits = 0

    const instructionMap: Map<string, { readonly totalGas: number; readonly hits: number }> =
        new Map()

    for (const line of lines) {
        if (line.info.$ !== "Covered") continue

        const lineGas = line.info.gasCosts.reduce((sum, gas) => sum + gas, 0)
        totalGas += lineGas
        totalHits += line.info.hits
        const trimmedLine = line.line.trim()
        const instructionName = trimmedLine.split(/\s+/)[0]
        if (instructionName !== undefined) {
            const current = instructionMap.get(instructionName) ?? {totalGas: 0, hits: 0}
            instructionMap.set(instructionName, {
                totalGas: current.totalGas + lineGas,
                hits: current.hits + line.info.hits,
            })
        }
    }

    const instructionStats: InstructionStat[] = [...instructionMap.entries()]
        .map(([name, stats]) => ({
            name,
            totalGas: stats.totalGas,
            totalHits: stats.hits,
            avgGas: stats.hits === 0 ? 0 : Math.round((stats.totalGas / stats.hits) * 100) / 100,
        }))
        .sort((a, b) => b.totalGas - a.totalGas)

    return {
        totalLines: totalExecutableLines,
        coveredLines,
        uncoveredLines,
        coveragePercentage,
        totalGas,
        totalHits,
        instructionStats,
    }
}

export const mergeCoverages = (...coverages: readonly Coverage[]): Coverage => {
    if (coverages.length === 0) {
        return {
            code: new Cell(),
            lines: [],
        };
    }

    let allLines: readonly Line[] = coverages[0]?.lines ?? [];
    for (const coverage of coverages.slice(1)) {
        allLines = mergeTwoLines(allLines, coverage.lines);
    }
    return {
        code: coverages[0]?.code ?? new Cell(),
        lines: allLines,
    };
};

export const mergeTwoLines = (
    first: readonly Line[],
    second: readonly Line[],
): readonly Line[] => {
    if (first.length !== second.length) return first;

    const result: Line[] = [...first];

    for (const [index, line] of second.entries()) {
        const prev = result[index];
        if (!prev) continue;

        if (prev.info.$ === "Uncovered" && line.info.$ === "Uncovered") {
            // nothing changes
            continue;
        }

        if (prev.info.$ === "Skipped" && line.info.$ === "Skipped") {
            // nothing changes
            continue;
        }

        if (prev.info.$ === "Uncovered" && line.info.$ === "Covered") {
            // replace it with new data
            result[index] = line;
        }

        if (prev.info.$ === "Covered" && line.info.$ === "Uncovered") {
            // nothing changes
            continue;
        }

        if (prev.info.$ === "Covered" && line.info.$ === "Covered") {
            result[index] = {
                ...prev,
                info: {
                    ...prev.info,
                    hits: prev.info.hits + line.info.hits,
                    gasCosts: [...prev.info.gasCosts, ...line.info.gasCosts],
                },
            };
        }
    }

    return result;
};

export const coverageToJson = (coverage: Coverage): string => {
    const lines = coverage.lines;
    return JSON.stringify({
        code: coverage.code.toBoc().toString("hex"),
        lines: lines.map((line, index) => {
            if (line.info.$ === "Covered") {
                return {
                    lineNumber: index,
                    line: line.line,
                    info: {
                        ...line.info,
                    },
                };
            }
            return {
                lineNumber: index,
                ...line,
            };
        }),
    });
};

export const coverageFromJson = (string: string): Coverage => {
    type CoverageJson = {
        readonly code: string;
        readonly lines: readonly Line[];
    };

    const data = JSON.parse(string) as CoverageJson;
    return {
        code: Cell.fromHex(data.code),
        lines: data.lines
    };
};
