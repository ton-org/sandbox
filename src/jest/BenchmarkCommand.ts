export interface BenchmarkCommandOption {
    label: string;
    doDiff: boolean;
}

export class BenchmarkCommand {
    readonly label?: string;
    readonly doDiff: boolean;

    constructor(option?: Partial<BenchmarkCommandOption>) {
        option = option || {};
        this.label = option?.label ?? process.env?.BENCH_NEW;
        this.doDiff = option?.doDiff ?? process.env?.BENCH_DIFF === 'true' ?? false;
    }

    get doBenchmark() {
        return this.doDiff || typeof this.label !== 'undefined';
    }
}
