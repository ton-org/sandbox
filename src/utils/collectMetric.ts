import { Address, beginCell, Cell, Contract } from '@ton/core';
import { Dictionary, DictionaryKeyTypes, TransactionComputePhase } from '@ton/core';
import { Blockchain, SendMessageResult } from '../blockchain/Blockchain';

export type MetricContext<T extends Contract> = {
    contract: T;
    methodName: string;
};

export type State = {
    code: Cell;
    data: Cell;
};

export type CellMetric = {
    cells: number;
    bits: number;
};

export type ComputePhaseMetric = {
    type: string;
    success?: boolean;
    gasUsed?: number;
    exitCode?: number;
    vmSteps?: number;
};

export type ActionPhaseMetric = {
    success?: boolean;
    totalActions?: number;
    skippedActions?: number;
    resultCode?: number;
    totalFwdFees?: number;
    totalActionFees?: number;
};

export type Metric = {
    testName?: string;
    address: string;
    contractName: string | null;
    methodName: string | null;
    opCode: number;
    computePhase: ComputePhaseMetric;
    actionPhase: ActionPhaseMetric;
    outMessages: CellMetric;
    state: CellMetric;
};

export type SnapshotMetric = {
    comment: string;
    createdAt: Date;
    items: Metric[];
};

const STORE_METRIC = Symbol.for('ton-sandbox-metric-store');

export function makeSnapshotMetric(comment: string, store: Metric[]): SnapshotMetric {
    const snapshot: SnapshotMetric = {
        comment,
        createdAt: new Date(),
        items: new Array<Metric>(),
    };
    const seen = new Set<string>();
    for (const metric of store) {
        const key = JSON.stringify(metric);
        if (seen.has(key)) continue;
        snapshot.items.push(metric);
        seen.add(key);
    }
    return snapshot;
}

export function getMetricStore(context: any = globalThis): Array<Metric> | undefined {
    return context[STORE_METRIC];
}

export function createMetricStore(context: any = globalThis): Array<Metric> {
    if (!Array.isArray(context[STORE_METRIC])) {
        context[STORE_METRIC] = new Array<Metric>();
    }
    return context[STORE_METRIC];
}

export function calcDictSize<K extends DictionaryKeyTypes, V>(dict: Dictionary<K, V>) {
    if (dict.size > 0) {
        return calcCellSize(beginCell().storeDict(dict).endCell().asSlice().loadRef());
    }
    return { cells: 0, bits: 0 };
}

export function calcCellSize(root: Cell, visited: Set<string> = new Set<string>()) {
    const hash = root.hash().toString('hex');
    if (visited.has(hash)) {
        return { cells: 0, bits: 0 };
    }
    visited.add(hash);
    let cells = 1;
    let bits = root.bits.length;
    for (const ref of root.refs) {
        const childRes = calcCellSize(ref, visited);
        cells += childRes.cells;
        bits += childRes.bits;
    }
    return { cells, bits };
}

export function calcStateSize(state: State): CellMetric {
    const codeSize = calcCellSize(state.code);
    const dataSize = calcCellSize(state.data);
    return {
        cells: codeSize.cells + dataSize.cells,
        bits: codeSize.bits + dataSize.bits,
    };
}

export function computePhase(phase: TransactionComputePhase): ComputePhaseMetric {
    if (phase.type === 'vm') {
        return {
            type: phase.type,
            success: phase.success,
            gasUsed: Number(phase.gasUsed),
            exitCode: phase.exitCode,
            vmSteps: phase.vmSteps,
        };
    }
    return {
        type: phase.type,
    };
}

export async function collectMetric<T extends Contract>(
    blockchain: Blockchain,
    ctx: MetricContext<T>,
    result: SendMessageResult,
) {
    const store = getMetricStore();
    if (!Array.isArray(store)) {
        return;
    }
    let state: CellMetric = { cells: 0, bits: 0 };
    if (ctx.contract.init && ctx.contract.init.code && ctx.contract.init.data) {
        state = calcStateSize({ code: ctx.contract.init.code, data: ctx.contract.init.data });
    } else {
        const account = (await blockchain.getContract(ctx.contract.address)).accountState;
        if (account && account.type === 'active' && account.state.code && account.state.data) {
            codeHash = `0x${account.state.code.hash().toString('hex')}`;
            state = calcStateSize({ code: account.state.code, data: account.state.data });
        }
    }
    let testName;
    if ((globalThis as any)['expect']) {
        testName = expect.getState().currentTestName;
    }
    let contractName: string | null = ctx.contract.constructor.name;
    let methodName: string | null = ctx.methodName;

    for (const tx of result.transactions) {
        if (tx.description.type !== 'generic') continue;
        const body = tx.inMessage?.info.type === 'internal' ? tx.inMessage?.body.beginParse() : undefined;
        const opCode = body && body.remainingBits >= 32 ? body.preloadUint(32) : 0;
        const address = Address.parse(`0:${tx.address.toString(16)}`);
        const metric: Metric = {
            testName,
            address: address.toString(),
            contractName,
            methodName,
            opCode,
            computePhase: computePhase(tx.description.computePhase),
            actionPhase: {
                success: tx.description.actionPhase?.success,
                totalActions: tx.description.actionPhase?.totalActions,
                skippedActions: tx.description.actionPhase?.skippedActions,
                resultCode: tx.description.actionPhase?.resultCode,
                totalActionFees: tx.description.actionPhase?.totalActions,
                totalFwdFees: Number(tx.description.actionPhase?.totalFwdFees ?? 0),
            },
            outMessages: calcDictSize(tx.outMessages),
            state,
        };
        store.push(metric);
        methodName = null;
        if (!address.equals(ctx.contract.address)) {
            contractName = ctx.contract.constructor.name;
        } else {
            contractName = null;
        }
    }
}
