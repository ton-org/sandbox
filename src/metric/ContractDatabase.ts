import { ContractABI, ABIReceiver, ABIType } from '@ton/core';
import { CodeHash, Metric, OpCode, isCodeHash } from './collectMetric';

type Condition = {
    codeHash: CodeHash;
    opCode: OpCode;
    receiver: 'internal' | 'external-in' | 'external-out';
};

export type ContractDataKey = CodeHash | string;

export type ContractData = Record<ContractDataKey, ContractABI | ContractDataKey>;

export class ContractDatabase {
    protected list: Map<ContractDataKey, ContractABI>;
    protected match: Map<ContractDataKey, ContractDataKey>;

    constructor(abiList: Map<ContractDataKey, ContractABI>, codeHashMatch: Map<ContractDataKey, ContractDataKey>) {
        this.list = abiList;
        this.match = codeHashMatch;
    }

    static from(data: ContractData): ContractDatabase {
        const list = new Map<ContractDataKey, ContractABI>();
        const match = new Map<ContractDataKey, ContractDataKey>();
        (Object.entries(data) as [ContractDataKey, ContractABI | ContractDataKey][]).forEach(([key, value]) => {
            if ((isCodeHash(key) && typeof value === 'string') || (isCodeHash(value) && typeof key === 'string')) {
                match.set(key, value);
            } else if (!isCodeHash(value) && typeof value !== 'string') {
                list.set(key, value);
            }
        });
        return new ContractDatabase(list, match);
    }

    get data(): ContractData {
        const out: ContractData = {};
        for (const [key, value] of this.match) {
            out[key] = value;
        }
        for (const [key, value] of this.list) {
            out[key] = value;
        }
        return out;
    }

    origin(needle: ContractDataKey) {
        return this.match.get(needle) || needle;
    }

    get(needle: ContractDataKey) {
        return this.list.get(this.origin(needle));
    }

    extract(metric: Metric) {
        const abiKeyNeedle = metric.contractName || metric.codeHash;
        if (!abiKeyNeedle) {
            return;
        }
        const codeHash = metric.codeHash;
        if (isCodeHash(codeHash) && abiKeyNeedle !== codeHash) {
            this.match.set(codeHash, abiKeyNeedle);
        }
        const abiKey = this.origin(abiKeyNeedle);
        const abi = this.list.get(abiKey) || ({} as ContractABI);

        if (!abi.receivers) {
            abi.receivers = [];
        }
        if (!abi.types) {
            abi.types = [];
        }
        const find = this.by(metric);
        if (!find.methodName) {
            if (!abi.name) {
                abi.name = metric.contractName || metric.codeHash;
            }
            if (metric.opCode !== '0x0') {
                abi.types.push({
                    name: metric.methodName || metric.opCode,
                    header: Number(metric.opCode),
                } as ABIType);
                abi.receivers.push({
                    receiver: metric.receiver == 'internal' ? 'internal' : 'external',
                    message: {
                        kind: 'typed',
                        type: metric.methodName || metric.opCode,
                    },
                } as ABIReceiver);
            }
        }
        this.list.set(abiKey, abi);
    }

    by(where: Partial<Condition>): Partial<Metric> {
        if (!where.codeHash) {
            return {};
        }
        const abi = this.get(where.codeHash);
        if (!abi) {
            return {};
        }
        const out: Partial<Metric> = {};
        out.contractName = abi.name ? abi.name : undefined;
        let abiType: ABIType | undefined;
        if (where.opCode) {
            for (const item of abi.types || []) {
                if (item.header && item.header === Number(where.opCode)) {
                    abiType = item;
                    break;
                }
            }
        }
        if (abiType) {
            const receiver = where.receiver ? (where.receiver == 'internal' ? 'internal' : 'external') : undefined;
            for (const item of abi.receivers || []) {
                if (receiver && receiver !== item.receiver) {
                    continue;
                }
                if (item.message.kind === 'typed' && item.message.type == abiType.name) {
                    out.methodName = item.message.type;
                    break;
                }
            }
        }
        return out;
    }
}
