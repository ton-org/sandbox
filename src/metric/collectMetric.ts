import { Address, beginCell, Cell, Contract, storeMessage, Message } from '@ton/core';
import { Dictionary, DictionaryKeyTypes, TransactionComputePhase } from '@ton/core';
import { Maybe } from '@ton/core/src/utils/maybe';

import { Blockchain, SendMessageResult } from '../blockchain/Blockchain';
import { ContractDatabase } from './ContractDatabase';

export type MetricContext<T extends Contract> = {
    contract: T;
    methodName: string;
};

type StateShort = {
    code: Cell;
    data: Cell;
};

export type CellMetric = {
    cells: number;
    bits: number;
};

export type ComputePhaseMetric = {
    type: string;
    success?: boolean;
    gasUsed?: number;
    exitCode?: number;
    vmSteps?: number;
};

export type ActionPhaseMetric = {
    success: boolean;
    totalActions: number;
    skippedActions: number;
    resultCode: number;
    totalFwdFees?: number;
    totalActionFees: number;
    totalMessageSize: CellMetric;
};

export type AddressFriendly = string;

export function isAddressFriendly(value: unknown): value is AddressFriendly {
    return typeof value === 'string' && Address.isFriendly(value);
}

export type ContractName = string;

export type ContractMethodName = string;

export type CodeHash = `0x${string}`;

export type OpCode = `0x${string}`;

export function isCodeHash(value: unknown): value is CodeHash {
    return typeof value === 'string' && value.length === 66 && /^0x[0-9a-fA-F]+$/.test(value);
}

export type StateMetric = {
    code: CellMetric;
    data: CellMetric;
};

export type Metric = {
    // the name of the current test (if available in Jest context)
    testName?: string;
    // address of contract
    address: AddressFriendly;
    // hex-formatted hash of contract code
    codeHash?: CodeHash;
    // total cells and bits usage of the contract's code and data
    state: StateMetric;
    contractName?: ContractName;
    methodName?: ContractMethodName;
    receiver?: 'internal' | 'external-in' | 'external-out';
    opCode: OpCode;
    // information from transaction phases
    execute: {
        compute: ComputePhaseMetric;
        action?: ActionPhaseMetric;
    };
    // total cells and bits usage of inbound and outbound messages
    message: {
        in: CellMetric;
        out: CellMetric;
    };
};

export type SnapshotMetric = {
    label: string;
    createdAt: Date;
    items: Metric[];
};

interface HasCreatedAt {
    createdAt: Date;
}

export function sortByCreatedAt(reverse = false) {
    return (a: HasCreatedAt, b: HasCreatedAt) => (a.createdAt.getTime() - b.createdAt.getTime()) * (reverse ? -1 : 1);
}

export type SnapshotMetricFile = {
    name: string;
    content: SnapshotMetric;
};

export type SnapshotMetricList = Record<string, SnapshotMetricFile>;

export type SnapshotMetricConfig = {
    label: string;
    contractExcludes: ContractName[];
    contractDatabase: ContractDatabase;
};

const STORE_METRIC = Symbol.for('ton-sandbox-metric-store');

export function makeSnapshotMetric(store: Metric[], config: Partial<SnapshotMetricConfig> = {}): SnapshotMetric {
    const label = config.label || 'current';
    const contractExcludes = config.contractExcludes || new Array<ContractName>();
    const contractDatabase = config.contractDatabase || ContractDatabase.from({});
    const snapshot: SnapshotMetric = {
        label,
        createdAt: new Date(),
        items: new Array<Metric>(),
    };
    // remove duplicates and extract ABI
    const seen = new Set<string>();
    for (const metric of store) {
        const key = JSON.stringify(metric);
        if (seen.has(key)) continue;
        snapshot.items.push(metric);
        seen.add(key);
        if (metric.codeHash) {
            contractDatabase.extract(metric);
        }
    }
    // ABI auto-mapping
    for (const item of snapshot.items) {
        const find = contractDatabase.by(item);
        if (!item.contractName && find.contractName) {
            item.contractName = find.contractName;
        }
        if (!item.methodName && find.methodName) {
            item.methodName = find.methodName;
        }
    }
    if (contractExcludes.length > 0) {
        snapshot.items = snapshot.items.filter(
            (it) => typeof it.contractName === 'undefined' || !contractExcludes.includes(it.contractName),
        );
    }
    return snapshot;
}

export function getMetricStore(context: any = globalThis): Array<Metric> | undefined {
    return context[STORE_METRIC];
}

export function createMetricStore(context: any = globalThis): Array<Metric> {
    if (!Array.isArray(context[STORE_METRIC])) {
        context[STORE_METRIC] = new Array<Metric>();
    }
    return context[STORE_METRIC];
}

export function resetMetricStore(context: any = globalThis): Array<Metric> {
    const store = getMetricStore(context);
    if (store) store.length = 0;
    return createMetricStore(context);
}

export function calcMessageSize(msg: Maybe<Message>) {
    if (msg) {
        return calcCellSize(beginCell().store(storeMessage(msg)).endCell());
    }
    return { cells: 0, bits: 0 };
}

export function calcDictSize<K extends DictionaryKeyTypes, V>(dict: Dictionary<K, V>) {
    if (dict.size > 0) {
        return calcCellSize(beginCell().storeDict(dict).endCell().asSlice().loadRef());
    }
    return { cells: 0, bits: 0 };
}

export function calcCellSize(root: Cell, visited: Set<string> = new Set<string>()) {
    const hash = root.hash().toString('hex');
    if (visited.has(hash)) {
        return { cells: 0, bits: 0 };
    }
    visited.add(hash);
    const out = {
        cells: 1,
        bits: root.bits.length,
    };
    for (const ref of root.refs) {
        const childRes = calcCellSize(ref, visited);
        out.cells += childRes.cells;
        out.bits += childRes.bits;
    }
    return out;
}

export function calcStateSize(state: StateShort): StateMetric {
    const codeSize = calcCellSize(state.code);
    const dataSize = calcCellSize(state.data);
    return {
        code: codeSize,
        data: dataSize,
    };
}

export function calcComputePhase(phase: TransactionComputePhase): ComputePhaseMetric {
    if (phase.type === 'vm') {
        return {
            type: phase.type,
            success: phase.success,
            gasUsed: Number(phase.gasUsed),
            exitCode: phase.exitCode,
            vmSteps: phase.vmSteps,
        };
    }
    return {
        type: phase.type,
    };
}

export enum OpCodeReserved {
    send = 0x0,
    notSupported = 0xffffffff,
    notAllowed = 0xfffffffe,
}

export async function collectMetric<T extends Contract>(
    blockchain: Blockchain,
    ctx: MetricContext<T>,
    result: SendMessageResult,
) {
    const store = getMetricStore();
    if (!Array.isArray(store)) {
        return;
    }
    let state: StateMetric = { data: { cells: 0, bits: 0 }, code: { cells: 0, bits: 0 } };
    let codeHash: CodeHash | undefined;
    if (ctx.contract.init && ctx.contract.init.code && ctx.contract.init.data) {
        codeHash = `0x${ctx.contract.init.code.hash().toString('hex')}`;
        state = calcStateSize({ code: ctx.contract.init.code, data: ctx.contract.init.data });
    } else {
        const account = (await blockchain.getContract(ctx.contract.address)).accountState;
        if (account && account.type === 'active' && account.state.code && account.state.data) {
            codeHash = `0x${account.state.code.hash().toString('hex')}`;
            state = calcStateSize({ code: account.state.code, data: account.state.data });
        }
    }
    let testName;
    if ((globalThis as any)['expect']) {
        // eslint-disable-next-line no-undef
        testName = expect.getState().currentTestName;
    }
    let contractName: ContractName | undefined = ctx.contract.constructor.name;
    let methodName: ContractMethodName | undefined = ctx.methodName;

    for (const tx of result.transactions) {
        if (tx.description.type !== 'generic') continue;
        const receiver = tx.inMessage?.info.type;
        const body = tx.inMessage?.body ? tx.inMessage.body.beginParse() : undefined;
        let opCode: OpCode = '0x0';
        if (receiver === 'internal') {
            opCode = `0x${(body && body.remainingBits >= 32 ? body.preloadUint(32) : 0).toString(16)}`;
        }
        if (!methodName && Object.values(OpCodeReserved).includes(Number(opCode))) {
            methodName = OpCodeReserved[Number(opCode)];
        }
        const address = Address.parseRaw(`0:${tx.address.toString(16).padStart(64, '0')}`);

        const { computePhase, actionPhase } = tx.description;
        const action: ActionPhaseMetric | undefined = actionPhase
            ? {
                  success: actionPhase.success,
                  totalActions: actionPhase.totalActions,
                  skippedActions: actionPhase.skippedActions,
                  resultCode: actionPhase.resultCode,
                  totalActionFees: actionPhase.totalActions,
                  totalFwdFees: actionPhase.totalFwdFees ? Number(actionPhase.totalFwdFees) : undefined,
                  totalMessageSize: {
                      cells: Number(actionPhase.totalMessageSize.cells),
                      bits: Number(actionPhase.totalMessageSize.bits),
                  },
              }
            : undefined;
        const compute = calcComputePhase(computePhase);
        const metric: Metric = {
            testName,
            address: address.toString(),
            codeHash,
            contractName,
            methodName,
            receiver,
            opCode,
            execute: {
                compute,
                action,
            },
            message: {
                in: calcMessageSize(tx.inMessage),
                out: calcDictSize(tx.outMessages),
            },
            state,
        };
        store.push(metric);
        methodName = undefined;
        if (!address.equals(ctx.contract.address)) {
            contractName = ctx.contract.constructor.name;
        } else {
            contractName = undefined;
        }
    }
}
