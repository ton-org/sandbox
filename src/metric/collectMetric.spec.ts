import '@ton/test-utils';
import { createMetricStore, getMetricStore, makeSnapshotMetric, resetMetricStore } from './collectMetric';
import { BenchmarkCommand } from '../jest/BenchmarkCommand';
import { ContractDatabase } from './ContractDatabase';
import { itIf, simpleCase } from './fixtures/data.fixture';

const bc = new BenchmarkCommand();

describe('collectMetric', () => {
    itIf(!bc.doBenchmark)('should not collect metric', async () => {
        let context: any = {};
        await simpleCase();
        expect(getMetricStore(context)).toEqual(undefined);
    });

    itIf(!bc.doBenchmark)('should be collect metric', async () => {
        const store = createMetricStore();
        resetMetricStore();
        expect(store.length).toEqual(0);
        await simpleCase();
        expect(store.length).toEqual(4);
        const contractDatabase = ContractDatabase.from({
            '0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912': 'TreasuryContract',
            TreasuryContract: {
                name: 'TreasuryContract',
                types: [{ name: 'foo', header: Number('0xdeadface'), fields: [] }],
                receivers: [
                    {
                        receiver: 'internal',
                        message: { kind: 'typed', type: 'foo' },
                    },
                ],
            },
        });
        const snapshot = makeSnapshotMetric(store, {
            label: 'foo',
            contractDatabase,
        });
        expect(snapshot.label).toEqual('foo');
        expect(snapshot.createdAt.getTime()).toBeLessThanOrEqual(new Date().getTime());
        expect(snapshot.items).toMatchSnapshot();
    });

    itIf(!bc.doBenchmark)('should be collect abi', async () => {
        const store = createMetricStore();
        resetMetricStore();
        expect(store.length).toEqual(0);
        await simpleCase();
        expect(store.length).toEqual(4);
        const contractDatabase = ContractDatabase.from({});
        const snapshot = makeSnapshotMetric(store, {
            label: 'foo',
            contractDatabase,
        });
        expect(snapshot.label).toEqual('foo');
        expect(snapshot.createdAt.getTime()).toBeLessThanOrEqual(new Date().getTime());
        expect(contractDatabase.data).toMatchSnapshot();
    });
});
