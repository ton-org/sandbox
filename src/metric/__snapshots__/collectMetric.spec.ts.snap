// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectMetric should be collect abi 1`] = `
{
  "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912": "TreasuryContract",
  "TreasuryContract": {
    "name": "TreasuryContract",
    "receivers": [
      {
        "message": {
          "kind": "typed",
          "type": "0xdeadface",
        },
        "receiver": "internal",
      },
      {
        "message": {
          "kind": "typed",
          "type": "notSupported",
        },
        "receiver": "internal",
      },
    ],
    "types": [
      {
        "header": 3735943886,
        "name": "0xdeadface",
      },
      {
        "header": 4294967295,
        "name": "notSupported",
      },
    ],
  },
}
`;

exports[`collectMetric should be collect metric 1`] = `
[
  {
    "address": "EQBiA46W-PQaaZZNFIDglnVknV9CR6J5hs81bSv70FwfNTrD",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "contractName": "TreasuryContract",
    "execute": {
      "action": {
        "resultCode": 0,
        "skippedActions": 0,
        "success": true,
        "totalActionFees": 1,
        "totalActions": 1,
        "totalFwdFees": 400000,
        "totalMessageSize": {
          "bits": 737,
          "cells": 1,
        },
      },
      "compute": {
        "exitCode": 0,
        "gasUsed": 1937,
        "success": true,
        "type": "vm",
        "vmSteps": 50,
      },
    },
    "message": {
      "in": {
        "bits": 998,
        "cells": 3,
      },
      "out": {
        "bits": 744,
        "cells": 2,
      },
    },
    "methodName": "send",
    "opCode": "0x0",
    "receiver": "external-in",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
  {
    "address": "EQBc3CG3NOeF3wwkBM8zjXrsWUhjLuN45LobSkZHXCR0jhvg",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "contractName": "TreasuryContract",
    "execute": {
      "action": {
        "resultCode": 0,
        "skippedActions": 0,
        "success": true,
        "totalActionFees": 0,
        "totalActions": 0,
        "totalFwdFees": undefined,
        "totalMessageSize": {
          "bits": 0,
          "cells": 0,
        },
      },
      "compute": {
        "exitCode": 0,
        "gasUsed": 309,
        "success": true,
        "type": "vm",
        "vmSteps": 5,
      },
    },
    "message": {
      "in": {
        "bits": 737,
        "cells": 1,
      },
      "out": {
        "bits": 0,
        "cells": 0,
      },
    },
    "methodName": "foo",
    "opCode": "0xdeadface",
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
  {
    "address": "EQBc3CG3NOeF3wwkBM8zjXrsWUhjLuN45LobSkZHXCR0jhvg",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "contractName": "TreasuryContract",
    "execute": {
      "action": {
        "resultCode": 0,
        "skippedActions": 0,
        "success": true,
        "totalActionFees": 1,
        "totalActions": 1,
        "totalFwdFees": 400000,
        "totalMessageSize": {
          "bits": 737,
          "cells": 1,
        },
      },
      "compute": {
        "exitCode": 0,
        "gasUsed": 1937,
        "success": true,
        "type": "vm",
        "vmSteps": 50,
      },
    },
    "message": {
      "in": {
        "bits": 998,
        "cells": 3,
      },
      "out": {
        "bits": 744,
        "cells": 2,
      },
    },
    "methodName": "send",
    "opCode": "0x0",
    "receiver": "external-in",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
  {
    "address": "EQBiA46W-PQaaZZNFIDglnVknV9CR6J5hs81bSv70FwfNTrD",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "contractName": "TreasuryContract",
    "execute": {
      "action": {
        "resultCode": 0,
        "skippedActions": 0,
        "success": true,
        "totalActionFees": 0,
        "totalActions": 0,
        "totalFwdFees": undefined,
        "totalMessageSize": {
          "bits": 0,
          "cells": 0,
        },
      },
      "compute": {
        "exitCode": 0,
        "gasUsed": 309,
        "success": true,
        "type": "vm",
        "vmSteps": 5,
      },
    },
    "message": {
      "in": {
        "bits": 737,
        "cells": 1,
      },
      "out": {
        "bits": 0,
        "cells": 0,
      },
    },
    "methodName": "notSupported",
    "opCode": "0xffffffff",
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
]
`;
